### DISCUSSION TITLE
Change dialogue depending on subsequent viewings.

### DISCUSSION THREAD

**Discussion by eisegesis-623** [nathanhoad/godot_dialogue_manager#402] [Ideas]
When using the "actionable" system to trigger dialogue, it's often desirable to show different text the second, or even third, time that you interact with the dialogue.
This functionality is of course _possible_ right now thanks to the ability to access variables in autoloads, but it's really verbose. The following code is my current solution (which I'm positive could be done much better).
```
~ label
if !MyGlobal.dialogue_dict.has("label"):
     MyGlobal.dialogue_dict["label"] = 0
if MyGlobal.dialogue_dict["label"] == 0
	MyGlobal.dialogue_dict["label"] += 1
	This is your first time looking at this.
	=> END
elif MyGlobal.dialogue_dict["label"] > 0
	MyGlobal.dialogue_dict["label"] += 1
	This is your second time looking at this.
	=> END
```
That's a LOT for two lines of dialogue (not to mention, it's kind of hard to follow). This sort of functionality is a lot simpler than "if character has X in inventory, then choose this dialogue" and doesn't need to persist beyond runtime
_(Having to add a new variable or key manually into the autoload every time you write a new dialogue label is redundant, so I'm doing it in the dialogue itself.)_

My proposal is to keep track of how many times a given label has been accessed in the DialogueManager itself. This way, all you need to do within dialogue is check to see how many times you've accessed it.

Here's a hypothetical example using `label(how_many_times_accessed)`.
```
if label(0):
	The first time looking at it.
elif label(1):
	The second time looking at it.
elif label(2,3):
	This will be the text the third and fourth times.
elif label(4):
	Since this is the last one, this is what the text will be every time after.
```

I can already see a lot of issues with this particular implementation. It doesn't really make sense how it would know to stop at `elif label(4)` with the if statement. And in the below example, you couldn't ever see the text "this is a different first time" if you accessed the dialogue before `SomeGlobal.variable` became `true`.
```
if SomeGlobal.variable == false:
	if label(0):
		The first time looking at it.
	elif label(1):
		The second time looking at it.
elif SomeGlobal.variable == true:
	if label(0):
		This is a different first time.
	elif label(1):
		This is a different second time.
```
Perhaps this could be solved by keeping track of how many times a particular dialogue _line_ has been accessed, rather than the label. Alternatively, using `=> another_label` to house the actual dialogue and `label(how_many_times_accessed)` checks could work.

I don't think anything I've suggested here is remotely close to a proper solution. Maybe handling this in the actionable itself is the best way to go, rather than in the dialogue. However, I do think there's a case to be made for this (or something similar) being a built-in feature. Thoughts?

**Comment by nathanhoad** [comment]
Dialogue Manager is stateless and tracking viewed lines would go against that. Tracking titles is something you would implement in your custom balloon by connecting to the `passed_title(title)` signal. Then you can pass `[self]` from the balloon in as the `extra_game_states` argument on [`get_next_dialogue_line(resource, title, extra_game_states)`](https://github.com/nathanhoad/godot_dialogue_manager/blob/main/docs/API.md#func-get_next_dialogue_lineresource-dialogueresource-key-string---extra_game_states-array---mutation_behaviour-mutationbehaviour--mutationbehaviourwait---dialogueline) - that way you have access to any methods you implement on your balloon in your dialogue.

**Comment by eisegesis-623** [comment]
@nathanhoad makes sense. Thanks so much for the reply!
